### DISCUSSION TITLE
Binding to custom class instance is always null

### DISCUSSION THREAD

**Discussion by simon-moisy** [AvaloniaUI/Avalonia#7513] [Q&A] [answered]
Hi all,

I am facing an issue with a simple app built with AvaloniaUI in MVVM. I have a custom control only defined in a C# class (draws things in the Render method). In this custom control I have a class instance property that I want to bind to my view model, but it is always null.
I am quite new to AvaloniaUI/MVVM approches so I probably did something wrong, but I can't see what.

Here's how I bind in the xaml : 
`<customControl:WaveFormControl WavePlayer="{Binding WavePlayer, Mode=TwoWay}" />`

And here is how I defined the WavePlayer and the DirectProperty in my control:
` public IWavePlayer? WavePlayer { get; set; }

    public static readonly DirectProperty<WaveFormControl, IWavePlayer?> WavePlayerProperty =
        AvaloniaProperty.RegisterDirect<WaveFormControl, IWavePlayer?>(
            nameof(WavePlayer), o => o.WavePlayer,
            (o, v) => o.WavePlayer = v);`

In my ViewModel I've declared my WavePlayer this way :
`
private IWavePlayer? _wavePlayer;
    public IWavePlayer? WavePlayer {
            get => _wavePlayer;        
            set => this.RaiseAndSetIfChanged(ref _wavePlayer, value);
    }`
Thanks in advance, any help would be much appreciated!

**Comment by ahopper** [comment]
it should be (o, v) => o.WavePlayer=v

**Reply by simon-moisy** [reply]
Sorry I made a mistake while copy pasting from my post on Gitter, I did indeed write it as you replied. I edit the main post, thanks!

**Comment by ahopper** [comment]
It could be an issue with your datacontext not being set or to what you expect

**Reply by simon-moisy** [reply]
Well in the same ViewModel I have two floats binding to sliders and a string to a property that belongs to the custom control (WaveFormControl), all three work without issue.
But anyway this is how the dataset is set :

In App.axaml.cs :
`                desktop.MainWindow = new MainWindow {
                    DataContext = new MainWindowViewModel(),
                };
`

In MainWindow.axaml :
`    <Design.DataContext>
        <vm:MainWindowViewModel/>
    </Design.DataContext>
`

That makes me wonder : is binding a standard type (string, floats) different from an instance of a custom class ?

**Comment by ahopper** [comment]
It is probably simplest to create a small repro project and post it somewhere, what you have posted looks ok to me so the problem could be elsewhere

**Comment by simon-moisy** [comment]
Alright I've pushed a small project [here](https://github.com/Tamaya31/BindingIssueMVVM) to demonstrate the issue. I am pretty sure I'm doing something wrong here but can't find out why.

Thanks in advance !

**Comment by ahopper** [comment]
In you repro you are binding the property to a null in your view model so you are setting it to null. If you set the dummyinstance to something in the viewModel it works.  Were you expecting the binding to go the other way?

**Comment by ahopper** [comment]
If you want a default value for the property it should be set in the property definition eg.
```
 public static readonly DirectProperty<DummyCustomControl, ADummyClass> ADummyInstanceProperty =
            AvaloniaProperty.RegisterDirect<DummyCustomControl, ADummyClass>(
                nameof(DummyCustomControl), o => o.ADummyInstance,
                (o, v) => o.ADummyInstance = v,new ADummyClass());
        ```

**Comment by simon-moisy** [comment]
Well, line 19 in the control I initialize ADummyInstance, and since I've added the `Mode=TwoWay` I was expecting the binding to work in both ways, isn't that what it's supposed to do ?

**Comment by ahopper** [comment]
you need to set the possible binding directions in the property definition as well as the default.

edit - this just sets the default binding mode not the possible ones.

**Comment by ahopper** [comment]
but two way binding takes its initial value from the source/viewmodel

**Comment by simon-moisy** [comment]
Thank you for your answer. To set the possible binding direction, do I do that through an attribute or something similar ? I can't find that info in the [binding documentation page](https://docs.avaloniaui.net/docs/data-binding/bindings#binding-modes)

**Comment by ahopper** [comment, marked as answer]
Sorry my comment about possible binding direction was wrong, that just sets the default.

your property should also be defined as 
```
ADummyClass _ADummyInstance;
        public ADummyClass ADummyInstance
        {
            get { return _ADummyInstance; }
            set { SetAndRaise(ADummyInstanceProperty, ref _ADummyInstance, value); }
        }
```
or else there is nothing to tell the system it has changed

**Comment by ahopper** [comment]
with the property defined like above and the binding changed to Mode=OneWayToSource in xaml I think it does what you expected

**Comment by simon-moisy** [comment]
Meanwhile your last message I actually did exactly that (the change with SetAndRaise), except changing the mode. And with the mode to OneWayToSource it works

Thanks a lot, I was struggling on that for a few days ! Hope this discussions helps others in the future too. Maybe the documentation can have a bit more information on that matter.
